lib/promise/group.rb

```diff
@@ -1,12 +1,15 @@
 class Promise
   class Group
+    include Promise::Observer
+
     attr_accessor :source
     attr_reader :promise
 
     def initialize(result_promise, inputs)
       @promise = result_promise
       @inputs = inputs
       @remaining = count_promises
+
       if @remaining.zero?
         promise.fulfill(inputs)
```

Comment:
@dylanahsmith This line here is problematic. `promise` gets fulfilled with the exact same value as passed into `Promise.all` which can lead to some really hard to understand bugs.

Imagine this:

```
input = [1, 2, 3]
Promise.all(input).then { |values|
  values << 4
}.then do |values|
  puts values #=> [1, 2, 3, 4]
  puts inputs #=> [1, 2, 3, 4]
end
```

vs

```
input = [Promise.resolve(1), 2, 3]
Promise.all(input).then { |values|
  values << 4
}.then do |values|
  puts values #=> [1, 2, 3, 4]
  puts inputs #=> [<Promise ...>, 2, 3]
end
```

Reply:
Do we want to prioritize the ability to modify the array result of Promise.all?  If we wanted to discourage that code, then an alternative would be to freeze the array created from mapping the input array.